downloader/src/main/kotlin/Downloader.kt

```diff
@@ -77,7 +80,39 @@ object Downloader {
 
         val exception = DownloadException("Download failed for '${pkg.id.toCoordinates()}'.")
 
-        // Try downloading from VCS.
+        var provenance: Provenance? = null
+        config.sourcesOrigin.forEach { origin ->
+            if (provenance == null) {
+                provenance = handleDownload(origin, pkg, outputDirectory, allowMovingRevisions, exception)
+            }
+        }
+
+        return provenance ?: throw exception
```

Comment:
How about

        return config.sourcesOrigin.asSequence().mapNotNull { origin ->
            handleDownload(origin, pkg, outputDirectory, allowMovingRevisions, exception)
        }.firstOrNull() ?: throw exception

Reply:
For safety, we might also want to add a `.distinct()` after `config.sourcesOrigin`.

Reply:
I don't think that `.asSequence()` is required or am I missing something?

Reply:
It's required to stop calling handleDownload() once the first download succeeded. In other words, the mapping is performed lazily this way.

Reply:
The hint about district() might have been bad, as it might internally convert to a set, which might change order. I cannot check this right now.

Reply:
> The hint about district() might have been bad, as it might internally convert to a set

It does convert it to a set.
How about using a `SortedSet` instead of a `List` for these values?

Reply:
Actually, `distinct()` should still work after all, as it's documented that "The elements in the resulting list are in the same order as they were in the source collection".